api/store/mongo/device_tags.go

```diff
@@ -35,19 +70,59 @@ func (s *Store) DevicePullTag(ctx context.Context, uid models.UID, tag string) e
 }
 
 func (s *Store) DeviceSetTags(ctx context.Context, uid models.UID, tags []string) (int64, int64, error) {
-	tag, err := s.db.Collection("devices").UpdateOne(ctx, bson.M{"uid": uid}, bson.M{"$set": bson.M{"tags": tags}})
+	tag, err := s.db.Collection("devices").
+		UpdateOne(ctx, bson.M{"uid": uid}, bson.M{"$set": bson.M{"tags": tags}})
 
 	return tag.MatchedCount, tag.ModifiedCount, FromMongoError(err)
 }
 
 func (s *Store) DeviceBulkRenameTag(ctx context.Context, tenant, currentTag, newTag string) (int64, error) {
-	res, err := s.db.Collection("devices").UpdateMany(ctx, bson.M{"tenant_id": tenant, "tags": currentTag}, bson.M{"$set": bson.M{"tags.$": newTag}})
+	res, err := s.db.Collection("devices").
+		UpdateMany(ctx, bson.M{"tenant_id": tenant, "tags": currentTag}, bson.M{"$set": bson.M{"tags.$": newTag}})
 
 	return res.ModifiedCount, FromMongoError(err)
 }
 
+func (s *Store) TagsBulkRenameTag(ctx context.Context, tenant, currentTag, newTag string) (int64, error) {
+	mongoSession, err := s.db.Client().StartSession()
+	if err != nil {
+		return 0, FromMongoError(err)
+	}
+	defer mongoSession.EndSession(ctx)
+
+	res := int64(0)
+
+	_, err2 := mongoSession.WithTransaction(ctx, func(sessCtx mongo.SessionContext) (interface{}, error) {
+		tag := new(models.Tags)
+
+		err := s.db.Collection("tags", options.Collection().SetWriteConcern(writeconcern.Majority())).
+			FindOne(sessCtx, bson.M{"tenant_id": tenant}).Decode(tag)
+		if err != nil {
+			return 0, FromMongoError(err)
+		}
+
+		res, err := s.TagsBulkDeleteTag(sessCtx, tenant, currentTag)
+		if err != nil {
+			return res, FromMongoError(err)
+		}
+
+		tag.Name = newTag
+
+		_, err2 := s.db.Collection("tags", options.Collection().SetWriteConcern(writeconcern.Majority())).
+			InsertOne(sessCtx, tag)
```

Comment:
This has to do with a way to simplify behavior in development as an atomic way and the old logic that no longer applies, since your observation, it really is better now to just update, thank you